src/blockchain_db/sqlite/db_sqlite.cpp

```diff
@@ -1720,4 +1692,174 @@ bool BlockchainSQLite::save_payments(
     }
     return true;
 }
+
+std::optional<uint64_t> BlockchainSQLite::fixup(bool recheck) {
+    auto tx = begin_tx();
+
+    auto with_commit = [&tx](std::optional<uint64_t> ret) {
+        if (tx)
+            tx->commit();
+        return ret;
+    };
+
+    auto prev_db_version = prepared_get<int>("PRAGMA user_version");
+    if (!recheck) {
+        if (prev_db_version >= FIXUP_DELAYED_PAYMENT_REWARDS)
+            return with_commit(std::nullopt);
+
+        db.exec("PRAGMA user_version = {}"_format(FIXUP_DELAYED_PAYMENT_REWARDS));
+    } else {
+        assert(prev_db_version >= FIXUP_DELAYED_PAYMENT_REWARDS);
+    }
+
+    if (nettype != cryptonote::network_type::MAINNET)
+        return with_commit(std::nullopt);
+
+    auto hf21_started = get_hard_fork_heights(nettype, hf::hf21_eth).first.value();
+    if (height < hf21_started)
+        // If we're before HF21 then there are no fixups to apply because we already have to rescan
+        // the entirety of HF21+ which should get everything right.
+        return with_commit(std::nullopt);
+
+    // If we select a straight sum of all amounts then we would overflow, so select the sum of
+    // values modulo a large (but not too large) prime as a checksum to verify the aggregate amount
+    // without having to resort to 128-bit math:
+    constexpr int64_t CSUM_MOD = 789012345678901;
+    struct fixup_record {
+        int64_t height;
+        int64_t checksum;
+    };
+    constexpr std::array<fixup_record, 4> RECORDS{{
+            {1'860'000, 212'642'223'336'227'860},
+            {1'870'000, 227'908'952'909'310'216},
+            {1'880'000, 19'447'225'076'325'721},
+            {1'890'000, 22'402'200'527'873'036},
+    }};
+
+    // If we can't find a way to do better, the fallback option is to require a rescan from HF21:
+    int64_t detach = hf21_started - 1;
+
+    // We are *always* going to force delete all HF21+ current and recent rows because we just don't
+    // know if they are correct and we need to rescan from one of the known-correct archive heights,
+    // above (or from our reproduced 1890k snapshot, if your node has an invalid 1890k value).
+    if (!recheck) {
+        db::exec_query(
+                db,
+                "DELETE FROM batched_payments_accrued_recent WHERE height >= ?",
+                static_cast<int64_t>(hf21_started));
+        if (height >= hf21_started)
+            db.exec("DELETE FROM batched_payments_accrued");
+    }
+
+    //
+    // We have snapshot data for 1890k, however, so if you were beyond 1890k but 1890k itself is
+    // incorrect, we'll start from there (and just drop any invalid pre-18
+
+    //
```

Comment:
This comment got chopped up and interrupted

Reply:
Whoops, I think it was edits that landed elsewhere; I've deleted it.